### PR TITLE
route: optimize virtual service route generation

### PR DESCRIPTION
**Pull request by ramaraochavali**
This PR fixes the following
- Moves `BuildHTTPRoutesForVirtualService` outside of for loop. I could not find a reason why it has to be loop.
- Changes `separateVSHostsAndServices` method to process non wildcard hosts first so that we take advantage of `serviceRegistry` map for them

### CODE REVIEW

**Comment by ramaraochavali** [comment]
our Route Generation Benchmark improvements before and after **first**  change
```
name                               old time/op    new time/op    delta
RouteGeneration/empty-12             4.01ms ±16%    2.68ms ±47%     ~     (p=0.056 n=5+5)
RouteGeneration/telemetry-12         4.60ms ±16%    3.02ms ±41%  -34.35%  (p=0.016 n=5+5)
RouteGeneration/virtualservice-12    4.44ms ±24%    2.73ms ±42%     ~     (p=0.056 n=5+5)

name                               old alloc/op   new alloc/op   delta
RouteGeneration/empty-12              709kB ± 0%     709kB ± 0%     ~     (p=0.841 n=5+5)
RouteGeneration/telemetry-12          765kB ± 0%     765kB ± 0%     ~     (p=0.198 n=5+5)
RouteGeneration/virtualservice-12     722kB ± 0%     723kB ± 0%     ~     (p=0.548 n=5+5)

name                               old allocs/op  new allocs/op  delta
RouteGeneration/empty-12              7.72k ± 0%     7.72k ± 0%     ~     (all equal)
RouteGeneration/telemetry-12          8.53k ± 0%     8.53k ± 0%     ~     (all equal)
RouteGeneration/virtualservice-12     7.97k ± 0%     7.97k ± 0%     ~     (all equal)
```

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 1]
```diff
@@ -214,11 +231,11 @@ func buildSidecarVirtualHostsForVirtualService(
 	}
 	meshGateway := map[string]bool{constants.IstioMeshGateway: true}
 	out := make([]VirtualHostWrapper, 0, len(serviceByPort))
+	routes, err := BuildHTTPRoutesForVirtualService(node, push, virtualService, serviceRegistry, listenPort, meshGateway)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:234]: good catch, seems a legacy implement

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 2]
```diff
@@ -158,21 +158,38 @@ func separateVSHostsAndServices(virtualService model.Config,
 	rule := virtualService.Spec.(*networking.VirtualService)
 	hosts := make([]string, 0)
 	servicesInVirtualService := make([]*model.Service, 0)
+	wchosts := make([]host.Name, 0)
+
+	// As a performance optimization, process non wildcard hosts first, so that they can be
+	// looked up directly in the service registry map.
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:164]: Not sure how much this will help, normally we will not have too many hosts

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:164]: There wont be may hosts but for non wildcard hosts we will loop through all the services which is not needed. See the benchmark results and also there was TODO for this also

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:164]: Understood

**Comment by howardjohn** [comment]
> -63.80%
Wow, great work

**Pull request merged**